src/resolve.rs

```diff
@@ -395,55 +416,198 @@ mod toml {
+            offset: 0,
+            source: ParseIndexError {
+                source: "invalid".parse::<usize>().unwrap_err(),
+            },
+        };
+        assert!(err.source().is_some());
+
+        let err = ResolveError::OutOfBounds {
+            offset: 0,
+            source: OutOfBoundsError {
+                index: 1,
+                length: 0,
+            },
+        };
+        assert!(err.source().is_some());
+
+        let err = ResolveError::NotFound { offset: 0 };
+        assert!(err.source().is_none());
+
+        let err = ResolveError::Unreachable { offset: 0 };
+        assert!(err.source().is_none());
     }
 
-    impl<'v, V> Test<'v, V>
-    where
-        V: Resolve<Value = V, Error = ResolveError>
-            + ResolveMut<Value = V, Error = ResolveError>
-            + Clone
-            + PartialEq
-            + fmt::Display
-            + fmt::Debug,
-    {
-        fn all(tests: impl IntoIterator<Item = Test<'v, V>>) {
-            tests.into_iter().enumerate().for_each(|(i, t)| t.run(i));
-        }
+    #[test]
+    fn resolve_error_display() {
+        let err = ResolveError::FailedToParseIndex {
+            offset: 0,
+            source: ParseIndexError {
+                source: "invalid".parse::<usize>().unwrap_err(),
+            },
+        };
+        assert_eq!(format!("{err}"), "failed to parse index at offset 0");
+
+        let err = ResolveError::OutOfBounds {
+            offset: 0,
+            source: OutOfBoundsError {
+                index: 1,
+                length: 0,
+            },
+        };
+        assert_eq!(format!("{err}"), "index at offset 0 out of bounds");
 
-        fn run(self, i: usize) {
-            _ = self;
-            let Test {
-                ptr,
-                data,
-                expected_result,
-            } = self;
-            let ptr = Pointer::from_static(ptr);
+        let err = ResolveError::NotFound { offset: 0 };
 
-            // cloning the data & expected_result to make comparison easier
-            let mut data = data.clone();
-            let expected_result = expected_result.cloned();
+        assert_eq!(format!("{err}"), "pointer starting at offset 0 not found");
 
-            // testing Resolve
-            let res = data.resolve(ptr).cloned();
-            assert_eq!(
-                &res, &expected_result,
-                "test #{i} failed:\n\nexpected\n{expected_result:#?}\n\nactual:\n{res:#?}",
-            );
+        let err = ResolveError::Unreachable { offset: 0 };
+        assert_eq!(
+            format!("{err}"),
+            "pointer starting at offset 0 is unreachable"
+        );
+    }
 
-            // testing ResolveMut
-            let res = data.resolve_mut(ptr).cloned();
-            assert_eq!(
-                &res, &expected_result,
-                "test #{i} failed:\n\nexpected\n{expected_result:#?}\n\nactual:\n{res:#?}",
-            );
-        }
+    #[test]
+    fn resolve_error_offset() {
```

Comment:
I think this one should use a different value to be more robust, 0 is more likely to be returned "by accident" than most values.

Also this is the sort of thing that would be more useful to check in a more end-to-end test, but no need to change it now.